alerting/alerter-war/src/main/java/org/hawkular/metrics/alerter/groups/GroupTriggerManager.java

```diff
@@ -0,0 +1,597 @@
+                } catch (Exception e) {
+                    log.error("Failed to fetch Conditions when refreshing Metrics Group Trigger " + mgt, e);
+                    return;
+                }
+                if (!dataIds.equals(conditionDataIds)) {
+                    log.warnf("Metrics Group Trigger dataId mismatch. In Tag: %s, In Conditions: %s. Skipping %s",
+                            dataIds, conditionDataIds, mgt);
+                    return;
+                }
+
+                // fetch the metric set for each dataId
+                Map<String, Set<Metric<?>>> dataIdMap = new HashMap<>();
+                for (String dataId : dataIds) {
+                    String tagQuery = tags.get(dataId);
+                    List<Metric<Object>> dataIdMetrics = metricsService
+                            .findMetricsWithFilters(mgt.getTenantId(), null, tagQuery)
+                            .toList()
+                            .toBlocking().firstOrDefault(Collections.emptyList());
+                    dataIdMap.put(dataId, new HashSet<>(dataIdMetrics));
+                }
+
+                // if there is no change in the dataIdMap since the last run we can stop here because the
+                // member set remains the same.
+                Integer hash = dataIdMap.hashCode();
+                if (hash.equals(dataIdMapHash)) {
+                    log.warnf("Metrics Group Trigger has no changes to member set, skipping %s", mgt);
+                    return;
+                }
+                dataIdMapHash = hash;
+
+                // Generate the member triggers.
+                // build a map of source->map<dataId, list<metricName>>
+                Map<List<String>, Map<String, List<String>>> sourceMap = generateSourceMap(tags.get(TAG_SOURCE_BY),
+                        dataIdMap);
+
+                // using the sourceMap generate the memberTriggers. Every tuple for every source is a valid member
+                // as long as it has a metricName for each dataId.
+                Set<GroupMemberInfo> members = generateMembers(mgt, dataIds, sourceMap);
+
+                // now fetch the existing members and synchronize
+                Collection<Trigger> existingMembers = definitionsService.getMemberTriggers(mgt.getTenantId(),
+                        mgt.getId(), true);
+                Set<String> memberIds = members.stream()
+                        .map(m -> m.getMemberId())
+                        .collect(Collectors.toSet());
+                Set<String> existingMemberIds = existingMembers.stream()
+                        .map(t -> t.getId())
+                        .collect(Collectors.toSet());
+                log.warnf("members: %s", members);
+                log.warnf("memberIds        : %s", memberIds);
+                log.warnf("existingMemberIds: %s", existingMemberIds);
+                for (GroupMemberInfo member : members) {
+                    if (existingMemberIds.contains(member.getMemberId())) {
```

Comment:
This wass another good observation, but I think this is OK, as you saw below, the member ID reflects the whole memberDataIdMapping by being the hashcode of the whole Map.